File: OmniVoice-Studio/backend/core/error_docs_map.py
"""Error class → docs URL mapping for the in-app deeplink button.

Used by the React ErrorBoundary's "Open docs for this error" button (via the
TypeScript mirror at `frontend/src/utils/errorDocsMap.ts`) and by the Phase 5
bug-reporter for "this error has a docs page" links.

The 5-class taxonomy below is the contract — Phase 5 reporter consumes it,
the TS map mirrors it, and `test_error_docs_map.test_keys_match_taxonomy`
locks the key set. To add a new class:

  1. Pick a stable ALL_CAPS key (the API will live forever).
  2. Add the entry here.
  3. Mirror it in `frontend/src/utils/errorDocsMap.ts`.
  4. Update the `test_keys_match_taxonomy` set and the TS-side keys-sync test.
"""
from __future__ import annotations

from core import links

_BASE = links.PROJECT_REPO_BLOB_MAIN

ERROR_DOCS: dict[str, str] = {
    "GATEKEEPER_QUARANTINE":       f"{_BASE}/docs/install/macos.md#gatekeeper-quarantine",
    "APPIMAGE_WEBKIT_WHITESCREEN": f"{_BASE}/docs/install/linux.md#appimage-white-screen-on-fedora-44--ubuntu-2404",
    "PKG_RESOURCES_MISSING":       f"{_BASE}/docs/install/troubleshooting.md#pkg_resources-missing",
    "HF_AUTH_FAILED":              f"{_BASE}/docs/setup/huggingface-token.md",
    # Issue #78 — pyannote/speaker-diarization-3.1 + pyannote/segmentation-3.0
    # are both gated on HuggingFace. A valid HF_TOKEN by itself isn't enough:
    # the user must also click "Agree and access repository" on both model
    # pages. `docs/features/diarization.md` walks through that flow in its
    # "License acceptance flow" section, so the deeplink targets that anchor
    # directly. Distinct from HF_AUTH_FAILED (which is the more general
    # token-missing-or-invalid case pointing at the token-setup doc).
    "PYANNOTE_LICENSE_REQUIRED":   f"{_BASE}/docs/features/diarization.md#license-acceptance-flow",
}

DEFAULT_DOCS: str = f"{_BASE}/docs/install/troubleshooting.md"


def lookup(error_class: str | None) -> str:
    """Return the docs URL for `error_class`, or DEFAULT_DOCS when the
    class is None / unknown."""
    return ERROR_DOCS.get(error_class or "", DEFAULT_DOCS)

File: OmniVoice-Studio/frontend/src/utils/errorDocsMap.ts
// MIRROR OF backend/core/error_docs_map.py — keep in sync.
// The Python test_keys_match_python_map / test_keys_match_taxonomy guards
// the 4-class taxonomy on the backend; the `_KEYS` array below is the
// TS-side anchor (the keys-sync test imports it and asserts equality).
//
// This `BASE` constant is the SECOND hardcoded URL drift site: the
// canonical Python-side resolver is `backend/core/links.py`
// (`PROJECT_REPO_BLOB_MAIN`). The TS half runs in the browser and can't
// read `pyproject.toml`, so it gets a hand-maintained mirror. Centralising
// the URL on the TS side is a deferred hardening item — for now we accept the
// drift risk and rely on the keys-sync test + threat-model T-02-01
// to bound the blast radius.

import { openExternal } from '../api/external';

const BASE = 'https://github.com/debpalash/VoiceStudio/blob/main';

export const ERROR_DOCS: Record<string, string> = {
  GATEKEEPER_QUARANTINE: `${BASE}/docs/install/macos.md#gatekeeper-quarantine`,
  APPIMAGE_WEBKIT_WHITESCREEN: `${BASE}/docs/install/linux.md#appimage-white-screen-on-fedora-44--ubuntu-2404`,
  PKG_RESOURCES_MISSING: `${BASE}/docs/install/troubleshooting.md#pkg_resources-missing`,
  HF_AUTH_FAILED: `${BASE}/docs/setup/huggingface-token.md`,
  // Issue #78 — pyannote gated-model license not accepted on HF.
  // Distinct from HF_AUTH_FAILED (which is a missing/invalid token).
  PYANNOTE_LICENSE_REQUIRED: `${BASE}/docs/features/diarization.md#license-acceptance-flow`,
};

export const DEFAULT_DOCS = `${BASE}/docs/install/troubleshooting.md`;

// Deep-link for the Dub tab's "needs install" translation-engine popover.
// Reuses BASE so it can't drift from the other GitHub-blob links above.
export const TRANSLATION_ENGINES_DOCS = `${BASE}/docs/dubbing/translation-engines.md#installing-optional-translation-engines-from-source-vs-packaged-build`;

// Locked taxonomy keys — Phase 5 bug reporter consumes this exact set.
// Adding a 6th class is a contract change; update the Python map at the
// same time (`backend/core/error_docs_map.py`).
export const ERROR_CLASS_KEYS = [
  'GATEKEEPER_QUARANTINE',
  'APPIMAGE_WEBKIT_WHITESCREEN',
  'PKG_RESOURCES_MISSING',
  'HF_AUTH_FAILED',
  'PYANNOTE_LICENSE_REQUIRED',
] as const;

export type ErrorClass = (typeof ERROR_CLASS_KEYS)[number];

/**
 * Heuristic error message → ErrorClass classifier. ErrorBoundary uses this
 * when the thrown Error doesn't carry an explicit `errorClass` property.
 */
export function classifyError(error: unknown): ErrorClass | null {
  const message =
    (error as { message?: string } | null | undefined)?.message ?? String(error ?? '');
  const lower = message.toLowerCase();
  if (/pkg_resources/.test(lower)) return 'PKG_RESOURCES_MISSING';
  // Issue #78 — pyannote license + diarization are diagnosed separately
  // from generic HF auth, since the fix instructions are different (click
  // "Agree" on the model page vs. set/refresh the token). Check this BEFORE
  // the HF_AUTH_FAILED branch so a message mentioning both "pyannote" and
  // "401" routes to the more specific deeplink.
  if (
    /pyannote/.test(lower) ||
    /\bgated\b/.test(lower) ||
    /speaker[- ]?diariz/.test(lower) ||
    /accept.*(license|terms|conditions)/.test(lower)
  ) {
    return 'PYANNOTE_LICENSE_REQUIRED';
  }
  if (/\b401\b/.test(lower) || /hfhub|hfhubhttp/.test(lower) || /unauthorized/.test(lower)) {
    return 'HF_AUTH_FAILED';
  }
  if (/webkit/.test(lower) || /white\s*screen/.test(lower)) {
    return 'APPIMAGE_WEBKIT_WHITESCREEN';
  }
  // Gatekeeper match: includes the literal "is damaged" macOS phrasing
  // (English + Chinese 已损坏 per issue #72). Lower-case test is safe;
  // '已损坏' is unaffected by lowercasing.
  if (
    /quarantine/.test(lower) ||
    /gatekeeper/.test(lower) ||
    /\bdamaged\b/.test(lower) ||
    /已损坏/.test(message || '')
  )
    return 'GATEKEEPER_QUARANTINE';
  return null;
}

export function urlFor(errorClass: ErrorClass | null | undefined): string {
  if (!errorClass) return DEFAULT_DOCS;
  return ERROR_DOCS[errorClass] ?? DEFAULT_DOCS;
}

/** Open the docs URL for `errorClass` in the user's default browser. */
export async function openDocsFor(
  errorClass: ErrorClass | string | null | undefined,
): Promise<void> {
  const url = ERROR_DOCS[errorClass as string] ?? DEFAULT_DOCS;
  await openExternal(url);
}
